### PR TITLE
Feature/pcp axis picking

### PR DESCRIPTION
**Pull request by Alboin**
I have added the ability to pick columns in the Parallel Coordinates Plot which are then sent to the brushing inport. Some properties had to be added such as axis width (hard to pick an axis that is 1 pixel in width), axis hover color and axis picking color.

Perhaps a BoolProperty should be added as well to the PCP to enable/disable axis picking/hovering, as it might not be a useful functionality for every user.

**What evineroment has the code been compiled and tested on?** 
* Windows 10
* Visual Studio 15.7.6
* QT Community 4.6.1
* CMake 3.11.2
* Python 3.6 (32-bit)

### CODE REVIEW

File: modules/plottinggl/glsl/pcp_axis.vert [hunk 1]
```diff
@@ -2,6 +2,11 @@
...
 
-void main(){
-    gl_Position = vec4(getPosWithSpacing(vec2(x,in_Vertex.y)),0,1);
-} 
\ No newline at end of file
+out vec3 pickColor;
+out vec2 texCoord;
+
+void main() {
+    pickColor = in_Normal.rgb;
```
Comment (ResearchDaniel) [inline modules/plottinggl/glsl/pcp_axis.vert]: Normal?

File: modules/plottinggl/src/processors/parallelcoordinates/parallelcoordinates.cpp [hunk 2]
```diff
@@ -81,6 +80,8 @@ ParallelCoordinates::ParallelCoordinates()
     , axisProperties_("axisProps_", "Axis")
     , colors_("colors", "Colors")
     , axisColor_("axisColor", "Axis Color", vec4(.6f, .6f, .6f, 1))
+    , axisHoverColor_("axisHoverColor", "Axis Hover Color", vec4(.3f, .3f, .3f, 1))
+    , axisSelectedColor_("axisSelectedColor", "Axis Selected Color", vec4(.8f, .0f, .0f, 1))
```
Comment (ResearchDaniel) [inline modules/plottinggl/src/processors/parallelcoordinates/parallelcoordinates.cpp]: Red color as default? How about a brighter color than the other axis

File: modules/plottinggl/include/modules/plottinggl/processors/parallelcoordinates/parallelcoordinates.h [hunk 3]
```diff
@@ -137,6 +140,8 @@ class IVW_MODULE_PLOTTINGGL_API ParallelCoordinates : public Processor {
     FloatProperty falllofPower_;
     FloatProperty lineWidth_;
     FloatProperty selectedLineWidth_;
+    FloatProperty axisWidth_;
+    FloatProperty selectedAxisWidth_;
```
Comment (ResearchDaniel) [inline modules/plottinggl/include/modules/plottinggl/processors/parallelcoordinates/parallelcoordinates.h]: Would it be possible to just have a thicker axis for picking , i.e. not visually? Then we could skip these properties and use a heuristic/hard coded values instead. 
Less is more :)

**Review by ResearchDaniel — APPROVED** [review]
Looks good to me! If possible, it would be good with some “automatic” variables so that the user does not need to specify them

File: modules/plottinggl/src/processors/parallelcoordinates/parallelcoordinates.cpp (hunk 2 above)
Comment (Alboin) [reply modules/plottinggl/src/processors/parallelcoordinates/parallelcoordinates.cpp]: Maybe fully white for selected, light gray for hover and dark gray by default?
pcp

File: modules/plottinggl/include/modules/plottinggl/processors/parallelcoordinates/parallelcoordinates.h (hunk 3 above)
Comment (Alboin) [reply modules/plottinggl/include/modules/plottinggl/processors/parallelcoordinates/parallelcoordinates.h]: We could. I just haven't figured it out in a nice way just yet ;)

Do we want the picking axis to be wider by a ratio or by a fix set of pixels?

File: modules/plottinggl/glsl/pcp_axis.vert (hunk 1 above)
Comment (Alboin) [reply modules/plottinggl/glsl/pcp_axis.vert]: I don't know. Copied it from the shader for the PCP lines.

File: modules/plottinggl/src/processors/parallelcoordinates/parallelcoordinates.cpp (hunk 2 above)
Comment (ResearchDaniel) [reply modules/plottinggl/src/processors/parallelcoordinates/parallelcoordinates.cpp]: Probably not fully white. A bit brighter than hover.

File: modules/plottinggl/include/modules/plottinggl/processors/parallelcoordinates/parallelcoordinates.h (hunk 3 above)
Comment (ResearchDaniel) [reply modules/plottinggl/include/modules/plottinggl/processors/parallelcoordinates/parallelcoordinates.h]: Can you check if the fragment Coord is more than one pixel from the center of the line and draw transparent color in that case?

Probably want it to be at least 3 pixels wide? If the axis is wider there is no need to extend the picking.

File: modules/plottinggl/glsl/pcp_axis.vert (hunk 1 above)
Comment (ResearchDaniel) [reply modules/plottinggl/glsl/pcp_axis.vert]: You are setting the pickColor uniform in cpp code. Should not use the normal

File: modules/plottinggl/src/processors/parallelcoordinates/parallelcoordinates.cpp (hunk 2 above)
Comment (Alboin) [reply modules/plottinggl/src/processors/parallelcoordinates/parallelcoordinates.cpp]: image

Better? 🥇

Comment (Alboin) [reply modules/plottinggl/src/processors/parallelcoordinates/parallelcoordinates.cpp]: The selected axis is also a little bit wider than the rest.

**Comment by Alboin** [comment]
Commit 254fe6b changes the lines of the PCP to (in my opinion) look better throughout the plot, except for at the start/end/joints of every line. They're now rendered equally wide, when previously the lines with a steep slope would be rendered thinner than those with no slope.

Old lines:
image

New lines:
image

**Comment by Alboin** [comment]
@petersteneteg Started to look at changing to the newer way of using shaders as you spoke about but the amount of work it would take for **me** would not be worth the results :)

So if you want to implement it you're welcome but for the GutBrainAxis-project it's not that important, the above commits will work just as fine.

**Pull request merged**